### ISSUE TITLE
Node filter for LocationIndex

### ISSUE THREAD

**Issue by michaz** [graphhopper/graphhopper#888]
I would like to have a variant of ``LocationTree.findClosest`` where I can pass a node filter instead of an ``EdgeFilter``. Like

```Java
public QueryResult findClosest(final double queryLat, final double queryLon, final Predicate<Integer> /* or something equivalent */ nodeFilter);
```

so that always

```Java
nodeFilter.test(findClosest(lat, lon, nodeFilter).getClosestNode()) == true
```

I tried to emulate this functionality with an ``EdgeFilter``, but to no avail, because edges are undirected.

**Comment by karussell** [comment]
Yes, the filter should be more generic. Highly related to #729 ... but I'm currently unsure how. Also there is a use case regarding map matching (findNClosest). Maybe a more generic filtering interface and returning a weighting list of QueryResults or something

**Comment by boldtrn** [comment]
@michaz could you provide a use case here? Right now nodes don't have a lot of information stored, basically Lat, Lon, Ele (if I am not mistaken).

**Comment by michaz** [comment]
@boldtrn .. and an ``additionalNodeField``, which in my case it is. The use case is public transit (branch pt), where one may want to find the nearest station.

Right now, I am using special loop edges to mark nodes and filter for them. Possible, but not as such elegant.

**Comment by michaz** [comment]
On the other hand, I could create _several_ indices: One for stations, one for road network nodes. But that is _also_ not easy, because the ``LocationIndex`` implementations are very tightly coupled to the ``Graph``, so it does not seem easy to create a ``LocationIndex`` and add some nodes and links to it.

**Comment by karussell** [comment]
> I have millions of inner nodes which should not be part of any index at all and just waste resources there

We could implement a filtering here, also necessary for road routing #322  

> But that is also not easy, because the LocationIndex implementations are very tightly coupled to the Graph

It stores just node IDs and uses the `Graph` interface to fetch more information, we could decouple this with another interface

**Comment by boldtrn** [comment]
@michaz Thanks, makes sense now. I was just wondering since in GH we usually don't have a lot of information per node.

**Comment by karussell** [comment]
> I was just wondering since in GH we usually don't have a lot of information per node

But we should definitely add something like this, e.g. making it easier to store OSM node IDs would be very helpful.

**Comment by karussell** [comment]
Another use case was mentioned in the forum: https://discuss.graphhopper.com/t/how-to-build-a-busflagencoder/1680
